### PR TITLE
Better logging

### PR DESCRIPTION
**Pull request by xamde**
Added logging and renamed sub-project directory. Improved warning reporting.

### CODE REVIEW

**Comment by xamde** [comment]
I don't understand why there are conflicts marked by GitHub.

File: pom.xml [hunk 1]
```diff
@@ -14,7 +14,7 @@
 
   <modules>
     <module>build-caching</module>
-    <module>j2cl-tools</module>
+    <module>j2cl-build</module>
```
Comment (niloc132) [inline pom.xml:17]: why rename/move this package?

File: j2cl-build/src/main/java/com/vertispan/j2cl/build/provided/ClosureTask.java [hunk 2]
```diff
@@ -153,6 +153,9 @@ public void execute(TaskOutput output) throws Exception {
             @Override
             public void finish(TaskOutput taskOutput) throws IOException {
                 Path webappDirectory = config.getWebappDirectory();
+                if(webappDirectory==null) {
+                    throw new IllegalArgumentException("No webapp directory defined");
```
Comment (niloc132) [inline j2cl-build/src/main/java/com/vertispan/j2cl/build/provided/ClosureTask.java:157]: I'm curious how you encountered this - also, I think this probably should be thrown from the config.getWebappDirectory method itself, rather than modify each final task to do a null check.

File: build-caching/src/main/java/com/vertispan/j2cl/build/TaskScheduler.java [hunk 3]
```diff
@@ -199,33 +206,35 @@ private void scheduleMoreWork(DiskCache.CacheResult cacheResult) {
...
-        System.out.println("starting final task " + taskDetails.getProject().getKey() + " " + taskDetails.getTaskFactory().getOutputType());
+        buildLog.info("starting final task " + taskDetails.getDebugName());
         long start = System.currentTimeMillis();
         try {
             ((TaskFactory.FinalOutputTask) taskDetails.getTask()).finish(new TaskOutput(cacheResult.outputDir()));
-            System.out.println(taskDetails.getProject().getKey() + " final task " + taskDetails.getTaskFactory().getOutputType() + " finished in " + (System.currentTimeMillis() - start) + "ms");
+            buildLog.info("Finished " + taskDetails.getDebugName() + " in " + (System.currentTimeMillis() - start) + "ms");
         } catch (Throwable t) {
-            System.out.println(taskDetails.getProject().getKey() + " final task " + taskDetails.getTaskFactory().getOutputType() + " failed in " + (System.currentTimeMillis() - start) + "ms");
+            buildLog.info("FAILED   " + taskDetails.getDebugName() + " in " + (System.currentTimeMillis() - start) + "ms");
```
Comment (niloc132) [inline build-caching/src/main/java/com/vertispan/j2cl/build/TaskScheduler.java:215]: this is probably an ERROR/WARN rather than INFO with a "FAILED" string?

File: build-caching/src/main/java/com/vertispan/j2cl/build/WatchService.java [hunk 4]
```diff
@@ -27,15 +28,18 @@
...
-    public WatchService(BuildService buildService, ScheduledExecutorService executorService) {
+    public WatchService(BuildService buildService, ScheduledExecutorService executorService, BuildLog log) {
         this.buildQueue = new BuildQueue(buildService);
         this.buildService = buildService;
         this.executorService = executorService;
+        this.buildLog =log;
     }
 
     public void watch(Map<Project, List<Path>> sourcePathsToWatch) throws IOException {
+        buildLog.info("Start watching "+sourcePathsToWatch);
```
Comment (niloc132) [inline build-caching/src/main/java/com/vertispan/j2cl/build/WatchService.java:42]: whitespace nit, spaces around the `+`

File: build-caching/src/main/java/com/vertispan/j2cl/build/TaskScheduler.java (hunk 3 above)
Comment (niloc132) [reply build-caching/src/main/java/com/vertispan/j2cl/build/TaskScheduler.java:215]: ...more likely a DEBUG/TRACE, depending on how long the task takes, but we can deal with that later.

File: build-caching/src/main/java/com/vertispan/j2cl/build/TaskScheduler.java [hunk 5]
```diff
@@ -94,6 +97,10 @@ public void onError(Throwable throwable) {
     }
 
     private void scheduleAvailableWork(Set<Input> ready, Map<Input, List<Input>> allInputs, Set<CollectedTaskInputs> remainingWork, BuildListener listener) {
+        if(remainingWork.size() == 1)
```
Comment (niloc132) [inline build-caching/src/main/java/com/vertispan/j2cl/build/TaskScheduler.java:100]: whitespace nit, after `if`, and add missing {}s

File: build-caching/src/main/java/com/vertispan/j2cl/build/PropertyTrackingConfig.java [hunk 6]
```diff
@@ -213,6 +213,14 @@ public String getLanguageOut() {
     @Override
     public Path getWebappDirectory() {
         // Note that this deliberately circumvents the hash building
-        return Paths.get(config.findNode("webappDirectory").readString());
+        ConfigValueProvider.ConfigNode configNode=config.findNode("webappDirectory");
+        if(configNode==null) {
+            throw new IllegalStateException("No 'webappDirectory' found");
+        }
+        String s = configNode.readString();
+        if(s==null) {
```
Comment (niloc132) [inline build-caching/src/main/java/com/vertispan/j2cl/build/PropertyTrackingConfig.java:221]: whitespace nits, after `if`, around `==`

File: build-caching/src/main/java/com/vertispan/j2cl/build/PropertyTrackingConfig.java [hunk 7]
```diff
@@ -213,6 +213,14 @@ public String getLanguageOut() {
     @Override
     public Path getWebappDirectory() {
         // Note that this deliberately circumvents the hash building
-        return Paths.get(config.findNode("webappDirectory").readString());
+        ConfigValueProvider.ConfigNode configNode=config.findNode("webappDirectory");
+        if(configNode==null) {
+            throw new IllegalStateException("No 'webappDirectory' found");
```
Comment (niloc132) [inline build-caching/src/main/java/com/vertispan/j2cl/build/PropertyTrackingConfig.java:218]: Ah good there is one here - can you remove the extra check from the final task?

File: Readme.md [hunk 8]
```diff
@@ -11,6 +11,9 @@ built from here:
 
 ------------------------
 
+# Usage
+A fully working sample project can be found at https://github.com/treblereel/j2cl-tests
```
Comment (niloc132) [inline Readme.md:15]: separate PR for unrelated changes please - though I have dmitrii's old archetype getting ready to merge, we'll want to call out both.

**Review by niloc132 — COMMENTED** [review]
Thanks for the PR! I'll figure out the merge issue if you can tidy up a bit.

--

Some potential followup for this review or a later one (I had been holding off on this work until I had a chance to incorporate all of these aspects as well):
I don't love putting the task name string in each time it is logged - would rather a delegate logger that wraps the existing logger and prefixes lines with the task.

This logger ideally then could be passed to the task itself while it works, to capture those changes with a helpful prefix. Each task's own logging should also probably be written to a files, so that warnings and such can be repeated on subsequent builds, instead of just silently continuing or saying "sorry, this failed last time, I can't build, clean to see what broke previously"

File: pom.xml (hunk 1 above)
Comment (xamde) [reply pom.xml:17]: The master project is already called "j2cl-tools" and having the same name in a sub-module seems weird. Furthermore, the pom.xml in the j2cl-tools subdirectory calls the artifact in there already "j2cl-build". Maven modules refer to local sub-directories (AFAIK). So this "rename" is more a clean-up.

Comment (niloc132) [reply pom.xml:17]: Good point - I'll reconsider naming a bit, this got thrashed a bit during that refactor. The goal is that this entire repo is no longer specific to maven (though it is built with maven), but also provides libraries that let other projects do their own j2cl building/watching (such as IDE tooling, a "java/j2cl fiddle" site, etc).

File: j2cl-build/src/main/java/com/vertispan/j2cl/build/provided/ClosureTask.java (hunk 2 above)
Comment (xamde) [reply j2cl-build/src/main/java/com/vertispan/j2cl/build/provided/ClosureTask.java:157]: I checked out an older sample project using <defaultWebappDirectory> and got really useless NPEs. Only with running maven plugin in debugger could I found out the issue. Moving the check to elsewhere sounds good. Not sure if I will figure that out.

Comment (niloc132) [reply j2cl-build/src/main/java/com/vertispan/j2cl/build/provided/ClosureTask.java:157]: You've actually already got a check in the right place, this should no longer be necessary.

File: build-caching/src/main/java/com/vertispan/j2cl/build/TaskScheduler.java (hunk 3 above)
Comment (xamde) [reply build-caching/src/main/java/com/vertispan/j2cl/build/TaskScheduler.java:215]: Maven has no "trace" level, so debug could be used. But since we throw the Throwable anyway, error seems fitting.

File: build-caching/src/main/java/com/vertispan/j2cl/build/PropertyTrackingConfig.java (hunk 6 above)
Comment (xamde) [reply build-caching/src/main/java/com/vertispan/j2cl/build/PropertyTrackingConfig.java:221]: What style checker is at work there? :-)

File: j2cl-build/src/main/java/com/vertispan/j2cl/build/provided/ClosureTask.java (hunk 2 above)
Comment (xamde) [reply j2cl-build/src/main/java/com/vertispan/j2cl/build/provided/ClosureTask.java:157]: Ok, got it

File: Readme.md (hunk 8 above)
Comment (xamde) [reply Readme.md:15]: This one confuses me, wasn't this one-liner already accepted as a separate PR before?

File: build-caching/src/main/java/com/vertispan/j2cl/build/TaskScheduler.java (hunk 3 above)
Comment (niloc132) [reply build-caching/src/main/java/com/vertispan/j2cl/build/TaskScheduler.java]: Maven doesn't, but other potential consumers of the library code might...

File: build-caching/src/main/java/com/vertispan/j2cl/build/PropertyTrackingConfig.java (hunk 6 above)
Comment (niloc132) [reply build-caching/src/main/java/com/vertispan/j2cl/build/PropertyTrackingConfig.java]: Absolutely a fair argument to make, but it might take longer to get your code merged.

Hey, want to make a patch to add a code formatter?

(Consistency is more important to me than "i like the formatter", but I've been very unhappy with "very consistent" formatters that take intent away, like chained stream calls each being on their own line, etc.)

File: Readme.md (hunk 8 above)
Comment (niloc132) [reply Readme.md:15]: It is possible you are not merged up to date?

Be sure you are using "main" and not "master", i'm using github's "change the main branch" workflow to change that. If you are using your IDE's git tooling, this might not be visible, I'm not sure.

Comment (xamde) [reply Readme.md:15]: It was this one https://github.com/Vertispan/j2clmavenplugin/pull/85

File: build-caching/src/main/java/com/vertispan/j2cl/build/PropertyTrackingConfig.java (hunk 6 above)
Comment (xamde) [reply build-caching/src/main/java/com/vertispan/j2cl/build/PropertyTrackingConfig.java]: I stopped having any opinions on code formatting :-) Consistency is really good to have better merges. I was just curious if you use some kind of tooling for automatic formatting which I could maybe replicate. **Manually copying the existing style is also fine.** I was just not paying attention, as I was used to some auto-formatting by IDE anyway. At work we use the google code formatter but have issues with latest Java 17 constructs. Nothing is perfect.

**Comment by niloc132** [comment]
Closed by #98
